Userland/Services/PreviewServer/PreviewProviders/Image.cpp

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2023, kleines Filmröllchen <filmroellchen@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Image.h"
+#include <LibCore/File.h>
+#include <LibCore/MimeData.h>
+#include <LibGfx/ImageDecoder.h>
+#include <LibGfx/Painter.h>
+
+REGISTER_PREVIEW_PROVIDER(Image)
+
+namespace PreviewServer::Providers {
+
+bool Image::can_generate_preview_for(String const& file)
+{
+    constexpr StringView image_mime_group = "image"sv;
+    // Ideally, detect image files without even opening the file.
+    if (Core::guess_mime_type_based_on_filename(file).starts_with(image_mime_group))
+        return true;
+
+    // Sniff 1KiB of the file's contents and try to guess the file type based on that.
+    auto maybe_open_file = Core::File::open(file, Core::File::OpenMode::Read);
+    if (maybe_open_file.is_error())
+        return false;
+    auto open_file = maybe_open_file.release_value();
+    auto maybe_first_kibibyte = ByteBuffer::create_uninitialized(KiB);
+    if (maybe_first_kibibyte.is_error())
+        return false;
+    auto first_kibibyte = maybe_first_kibibyte.release_value();
+    auto read_result = open_file->read_until_filled(first_kibibyte);
+    if (read_result.is_error())
+        return false;
+
+    if (Core::guess_mime_type_based_on_sniffed_bytes(first_kibibyte)->starts_with(image_mime_group))
+        return true;
+    return false;
+}
+
+CacheEntry Image::generate_preview(String const& file)
+{
+    dbgln_if(PREVIEW_SERVER_DEBUG, "Generating image preview for {}", file);
+    auto open_file = TRY(Core::File::open(file, Core::File::OpenMode::Read));
+    auto image_data = TRY(open_file->read_until_eof());
+    auto maybe_image = TRY(image_decoder_client())->decode_image(image_data);
+    if (!maybe_image.has_value())
+        return Error::from_string_view("Image decoding failed"sv);
+    auto image_descriptor = maybe_image->frames[0];
+    if (image_descriptor.bitmap.is_null())
+        return Error::from_string_view("Reading first frame of image failed"sv);
+    auto image = image_descriptor.bitmap.release_nonnull();
+
+    auto preview = TRY(Gfx::Bitmap::create_shareable(Gfx::BitmapFormat::BGRA8888, { preview_size, preview_size }));
+
+    double scale = min(32 / (double)image->width(), 32 / (double)image->height());
```

Comment:
(itʼd be nice to make the size constant or parameterised and avoid referring to the specific size in the code; also: in comments.)